source/iterator/cdc.go

```diff
@@ -0,0 +1,289 @@
+		return nil, fmt.Errorf("setup vstream: %w", err)
+	}
+
+	go cdc.listen(ctx)
+
+	return cdc, nil
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+func (c *CDC) HasNext(ctx context.Context) (bool, error) {
+	return len(c.records) > 0, nil
+}
+
+// Next returns the next record.
+func (c *CDC) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+
+	case err := <-c.errCh:
+		return sdk.Record{}, err
+
+	case record := <-c.records:
+		return record, nil
+	}
+}
+
+// Stop closes the underlying db connection.
+func (c *CDC) Stop(ctx context.Context) error {
+	if c.conn != nil {
+		c.conn.Close()
+	}
+
+	return nil
+}
+
+// setupVStream opens a connection to a vtgate and create a VStream reader.
+// The method returns the connection, the VStream reader and a gtid.
+func (c *CDC) setupVStream(ctx context.Context, params CDCParams) error {
+	shardGtid := &binlogdata.ShardGtid{
+		Keyspace: params.Keyspace,
+		// if the Shard is empty, we'll get rows from all the available shards.
+		Shard: "",
+		Gtid:  c.position.Gtid,
+	}
+
+	vgtid := &binlogdata.VGtid{
+		ShardGtids: []*binlogdata.ShardGtid{shardGtid},
+	}
+
+	ruleFilter, err := c.constructRuleFilter(params.Table, params.OrderingColumn, params.Columns)
+	if err != nil {
+		return fmt.Errorf("construct rule filter: %w", err)
+	}
+
+	filter := &binlogdata.Filter{
+		Rules: []*binlogdata.Rule{{
+			Match:  params.Table,
+			Filter: ruleFilter,
+		}},
+	}
+
+	conn, err := vtgateconn.Dial(ctx, params.Address)
+	if err != nil {
+		return fmt.Errorf("vtgateconn dial: %w", err)
+	}
+	c.conn = conn
+
+	tabletType := topodata.TabletType(topodata.TabletType_value[params.TabletType])
+
+	reader, err := conn.VStream(ctx, tabletType, vgtid, filter, &vtgate.VStreamFlags{
+		MinimizeSkew: false,
+	})
+	if err != nil {
+		return fmt.Errorf("create vstream reader: %w", err)
+	}
+	c.reader = reader
+
+	return nil
+}
+
+// constructRuleFilter constructs an SQL query for the binlogdata.Filter.Rules.
+func (c *CDC) constructRuleFilter(table, orderingColumn string, columns []string) (string, error) {
+	selectBuilder := sqlbuilder.NewSelectBuilder()
+
+	if len(columns) > 0 {
+		for i, column := range columns {
+			columns[i] = sqlbuilder.MySQL.Quote(column)
+		}
+
+		selectBuilder.Select(columns...)
+	} else {
+		selectBuilder.Select("*")
+	}
+
+	selectBuilder.From(table).OrderBy(orderingColumn)
+
+	query, err := sqlbuilder.MySQL.Interpolate(selectBuilder.BuildWithFlavor(sqlbuilder.MySQL))
+	if err != nil {
+		return "", fmt.Errorf("interpolate arguments to SQL: %w", err)
+	}
+
+	return query, nil
+}
+
+// listen listens for VStream events.
+// If the VStream encountered an error the method will send it to the errCh channel.
+// All the data are sent to the records channel.
+func (c *CDC) listen(ctx context.Context) {
+	for {
+		events, err := c.reader.Recv()
```

Comment:
Actually no, we couldn't. Only if we would start from the shard specified as an empty string `""`, but this case is not possible because by default we use `"current"` which means that we will receive events from the current point of time